processing/src/main/java/io/druid/segment/IndexIO.java

```diff
@@ -251,6 +245,78 @@ public static boolean convertSegment(File toConvert, File converted, IndexSpec i
     public MMappedIndex mapDir(File inDir) throws IOException;
   }
 
+  public static class ValidationException extends IOException{
+    public ValidationException(String formatText, Object... arguments)
+    {
+      super(String.format(formatText, arguments));
+    }
+
+    public ValidationException(Throwable cause, String formatText, Object... arguments)
+    {
+      super(String.format(formatText, arguments), cause);
+    }
+  }
+
+  public static void validateRowValues(
+      Rowboat rb1,
+      IndexableAdapter adapter1,
+      Rowboat rb2,
+      IndexableAdapter adapter2
+  ) throws ValidationException
+  {
+    final int[][] dims1 = rb1.getDims();
+    final int[][] dims2 = rb2.getDims();
+    if (dims1.length != dims2.length) {
+      throw new ValidationException(
+          "Dim lengths not equal %s vs %s",
+          Arrays.deepToString(dims1),
+          Arrays.deepToString(dims2)
+      );
+    }
+    final Indexed<String> dim1Names = adapter1.getDimensionNames();
+    final Indexed<String> dim2Names = adapter2.getDimensionNames();
+    for (int i = 0; i < dims1.length; ++i) {
+      final int[] dim1Vals = dims1[i];
+      final int[] dim2Vals = dims2[i];
+      final String dim1Name = dim1Names.get(i);
+      final String dim2Name = dim2Names.get(i);
+
+      if (dim1Vals.length != dim2Vals.length) {
+        throw new ValidationException(
+            "Dim [%s] value lengths not equal %s vs %s",
+            dim1Name,
+            Arrays.deepToString(dims1),
+            Arrays.deepToString(dims2)
+        );
+      }
+
+      // Compare dim values
+      final Indexed<String> dim1ValNames = adapter1.getDimValueLookup(dim1Name);
+      final Indexed<String> dim2ValNames = adapter2.getDimValueLookup(dim2Name);
+      for (int j = 0; j < dim1Vals.length; ++i) {
+        final int dIdex1 = dims1[i][j];
+        final int dIdex2 = dims2[i][j];
+        if (dIdex1 == dIdex2) {
+          continue;
+        }
+
+        final String dim1ValName = dim1ValNames.get(dIdex1);
+        final String dim2ValName = dim2ValNames.get(dIdex2);
+        if ((dim1ValName == null) || (dim2ValName == null)) {
+          if ((dim1ValName == null) && (dim2ValName == null)) {
+            continue;
+          } else {
+            throw new ValidationException("Dim [%s] value not equal %s vs %s", dim1Name, dim1ValName, dim2ValName);
```

Comment:
I did it this way because I originally had a bunch of errors with "Validation exception" at the beginning. And realized it would end up being cleaner if "Validation exception" was the exception type instead of in the message.